src/authentication.rs

```diff
@@ -0,0 +1,158 @@
+use nex_rs::client::{ClientConnection};
+use nex_rs::nex_types::StructureInterface;
+use nex_rs::packet::{Packet, PacketV1};
+use nex_rs::stream::{StreamIn};
+use no_std_io::{Cursor, Reader, StreamReader};
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+pub const AUTHENTICATION_PROTOCOL_ID: u8 = 0xA;
+
+#[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum AuthenticationMethod {
+    Login = 0x1,
+    LoginEx = 0x2,
+    RequestTicket = 0x3,
+    GetPID = 0x4,
+    GetName = 0x5,
+    LoginWithParam = 0x6,
+}
+
+#[derive(Default)]
+pub struct NintendoLoginData(String);
+
+#[derive(Default)]
+pub struct AuthenticationInfo {
+    token: String,
+    ngs_version: u32,
+    token_type: u8,
+    server_version: u32,
+}
+
+impl AuthenticationInfo {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl StructureInterface for AuthenticationInfo {
+    fn extract_from_stream<T: Reader>(&mut self, stream: &mut StreamIn<T>) -> Result<(), &'static str> {
+        let token = stream.read_string();
+        if stream.get_slice()[stream.get_index()..].len() < 9 {
+            return Err("[AuthenticationInfo::extract_from_stream] Data size too small");
+        }
+
+        self.token = token;
+        self.token_type = stream.read_stream()
+            .map_err(|_| "[AuthenticationInfo::extract_from_stream] Failed to read token type")?;
+        self.ngs_version = stream.read_stream_le()
+            .map_err(|_| "[AuthenticationInfo::extract_from_stream] Failed to read NGS version")?;
+        self.server_version = stream.read_stream_le()
+            .map_err(|_| "[AuthenticationInfo::extract_from_stream] Failed to read server version")?;
+
+        Ok(())
+    }
+}
+
+pub trait AuthenticationProtocol {
+    fn login(&self, _client: &mut ClientConnection, _call_id: u32, _username: String) -> Result<(), &'static str> { Ok(()) }
+    fn login_ex(&self, _client: &mut ClientConnection, _call_id: u32, _username: String, _authentication_info: Option<AuthenticationInfo>) -> Result<(), &'static str> { Ok(()) }
+    fn request_ticket(&self, _client: &mut ClientConnection, _call_id: u32, _user_pid: u32, _server_pid: u32) -> Result<(), &'static str> { Ok(()) }
+    fn get_pid(&self, _client: &mut ClientConnection, _call_id: u32, _username: String) -> Result<(), &'static str> { Ok(()) }
+    fn get_name(&self, _client: &mut ClientConnection, _call_id: u32, _user_pid: u32) -> Result<(), &'static str> { Ok(()) }
+    fn login_with_param(&self, _: &mut ClientConnection, _call_id: u32) -> Result<(), &'static str> { Ok(()) }
```

Comment:
Instead of default implementations that return `Ok`, what are thoughts on leaving these unimplemented so users are required to implement them?
```suggestion
    fn login(&self, client: &mut ClientConnection, call_id: u32, username: String) -> Result<(), &'static str>;
    fn login_ex(&self, client: &mut ClientConnection, call_id: u32, username: String, authentication_info: Option<AuthenticationInfo>) -> Result<(), &'static str>;
    fn request_ticket(&self, client: &mut ClientConnection, call_id: u32, user_pid: u32, server_pid: u32) -> Result<(), &'static str>;
    fn get_pid(&self, client: &mut ClientConnection, call_id: u32, username: String) -> Result<(), &'static str>;
    fn get_name(&self, client: &mut ClientConnection, call_id: u32, user_pid: u32) -> Result<(), &'static str>;
    fn login_with_param(&self, client: &mut ClientConnection, call_id: u32) -> Result<(), &'static str>;
```